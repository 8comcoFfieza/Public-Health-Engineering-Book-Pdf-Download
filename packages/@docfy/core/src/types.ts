import { Node } from 'unist';
import { Processor, Plugin } from 'unified';

export interface Page {
  source: string;
  ast: Node;
  markdown: string;
  rendered: string;
  metadata: {
    title?: string;
    order?: number;
    package?: string;
    category?: string;
    url?: string;
  };
  demos?: Page[];
}

export interface Context {
  root: string;
  remark: Processor;
  pages: Page[];
}

interface SourceSettings {
  pattern: string;
  urlPrefix?: string;
  ignore?: string[];
}

export interface Options {
  root: string;
  sources: SourceSettings[];
  remarkPlugins?: Plugin[];
}
