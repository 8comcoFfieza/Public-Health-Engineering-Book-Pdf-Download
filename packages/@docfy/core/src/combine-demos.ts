import path from 'path';
import { Page, Context } from './types';

/*
 * Finds the index of the owner for a given demo file.
 *
 * Demos should be located in a folder named `demo/` or `something-demo/`.
 *
 * eg.
 * Below the owner would be the button/index.md:
 * components/button/
 *   - index.md
 *   - demo/
 *     - demo1.md
 *     - demo2.md
 *
 * Below the owner would be the button.md:
 * components/
 *   - button.md
 *   - button-demo/
 *     - demo1.md
 *     - demo2.md
 */
function findDemoOwner(contents: Page[], demoSource: string): number {
  const folder = path.basename(path.dirname(demoSource));

  let parentName = folder.replace('-demo', '');
  if (parentName === 'demo') {
    parentName = path.basename(path.dirname(path.dirname(demoSource)));
  }

  return contents.findIndex((item: Page): boolean => {
    const file = path.parse(path.basename(item.source));
    return (
      file.name === parentName ||
      (file.name === 'index' &&
        path.basename(path.dirname(item.source)) === parentName)
    );
  });
}

export default function CombineDemos(context: Context): Context {
  context.pages.forEach((item: Page, index: number): void => {
    const folder = path.basename(path.dirname(item.source));

    if (folder.match(/demo$/)) {
      const parent =
        context.pages[findDemoOwner(context.pages, item.source)];

      if (parent) {
        if (!parent.demos) {
          parent.demos = [item];
        } else {
          parent.demos.push(item);
        }

        context.pages.splice(index, 1);
      }
    }
  });

  return context;
}
