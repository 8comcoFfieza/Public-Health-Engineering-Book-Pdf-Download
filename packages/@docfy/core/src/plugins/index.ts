export { fixUrls } from './fix-urls';
export { combineDemos } from './combine-demos';
