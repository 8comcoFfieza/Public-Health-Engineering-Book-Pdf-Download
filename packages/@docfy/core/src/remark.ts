import unified from 'unified';
import parse from 'remark-parse';
import frontmatter from 'remark-frontmatter';
import slug from 'remark-slug';
import html from 'remark-html';
import stringify from 'remark-stringify';
import normalizeHeadings from 'remark-normalize-headings';
import { Processor } from 'unified';
import { Options } from './types';

export function createRemark(plugins?: Options['remarkPlugins']): Processor {
  const stack = unified()
    .use(parse)
    .use(normalizeHeadings)
    .use(slug)
    .use(stringify)
    .use(frontmatter, [{ type: 'yaml', marker: '-' }]);

  if (plugins && plugins.length > 0) {
    plugins.forEach((fn) => {
      if (Array.isArray(fn)) {
        stack.use(...fn);
      } else {
        stack.use(fn);
      }
    });
  }

  return stack.use(html);
}
