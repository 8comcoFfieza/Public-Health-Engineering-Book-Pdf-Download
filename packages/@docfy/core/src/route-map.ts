import { Content } from './types';

export default function routeMap(content: Content[], prefix = ''): string[] {
  const routes: string[] = [];

  // intro, bla, category: [bla, bla], package: [bla, bla, category]

  content.forEach((item) => {
    if (item.metadata.url) {
      routes.push(prefix + item.metadata.url);
    }
  });

  return routes;
}
