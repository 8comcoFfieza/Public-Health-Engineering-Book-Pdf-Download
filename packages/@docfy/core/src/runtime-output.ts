import { Page, PageContent, RuntimeOutput, NestedRuntimeOutput } from './types';

function findChild(
  node: NestedRuntimeOutput,
  name: string
): NestedRuntimeOutput | undefined {
  return node.children.find((item) => {
    return item.name === name;
  });
}

function sortByOrder(pages: Page[]): Page[] {
  return pages.sort((a, b) => {
    const aOrder =
      typeof a.metadata.order !== 'undefined' ? Number(a.metadata.order) : 998;
    const bOrder =
      typeof b.metadata.order !== 'undefined' ? Number(b.metadata.order) : 999;
    return aOrder - bOrder;
  });
}

export function genereateFlatOutput(pageContents: PageContent[]): Page[] {
  const pages: Page[] = pageContents.map((page) => {
    const { url, headings, title, source, metadata } = page;

    return {
      url,
      headings,
      title,
      source,
      metadata
    };
  });

  return sortByOrder(pages);
}

export function generateNestedOutput(
  pages: Page[],
  labels: Record<string, string> = {},
  existingObj?: NestedRuntimeOutput
): NestedRuntimeOutput {
  const node: NestedRuntimeOutput = existingObj || {
    name: '/',
    label: labels['/'] || '/',
    pages: [],
    children: []
  };

  pages.forEach((item): void => {
    let url = (item.metadata.relativeUrl as string) || item.url;
    url = url[0] === '/' ? url.substring(1) : url;
    const urlParts = url.split('/');

    if (urlParts.length === 1) {
      item.metadata.relativeUrl = urlParts[0];
      node.pages.push(item);
    } else {
      const name = urlParts.shift();

      if (typeof name === 'string') {
        let child = findChild(node, name);

        if (!child) {
          child = {
            name: name,
            label: labels[name] || name,
            pages: [],
            children: []
          };
          node.children.push(child);
        }

        item.metadata.relativeUrl = urlParts.join('/');
        generateNestedOutput([item], labels, child);

        sortByOrder(child.pages);
      }
    }
  });

  sortByOrder(node.pages);
  return node;
}

export function generateRuntimeOutput(
  pageContents: PageContent[],
  labels: Record<string, string> = {}
): RuntimeOutput {
  const flat = genereateFlatOutput(pageContents);
  const nested = generateNestedOutput(flat, labels);

  return { flat, nested };
}
