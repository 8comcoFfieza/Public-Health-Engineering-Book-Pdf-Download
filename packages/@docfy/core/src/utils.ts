import path from 'path';
import visit from 'unist-util-visit';
import { Node } from 'unist';
import { Page } from './types';
import toString from 'mdast-util-to-string';
import Slugger from 'github-slugger';
import YAML from 'yaml';
import url from 'url';

const slug = Slugger.slug;

export function generateManualUrl(
  source: string,
  meta: Page['metadata'],
  prefix?: string,
  suffix?: string
): string {
  const parts: string[] = [''];
  if (prefix) {
    parts.push(prefix);
  }
  if (meta.package) {
    parts.push(slug(meta.package));
  }
  if (meta.category) {
    parts.push(slug(meta.category));
  }
  let fileName = path.parse(path.basename(source)).name;
  if (fileName === 'index') {
    fileName = path.basename(path.dirname(source));
  }
  parts.push(`${fileName}${suffix || ''}`);
  return parts.join('/');
}

export function generateAutoUrl(
  source: string,
  prefix?: string,
  suffix?: string
): string {
  const parts: string[] = [''];
  if (prefix) {
    parts.push(prefix);
  }
  parts.push(...source.split(path.sep));

  let fileName = path.parse(parts.pop() as string).name;
  if (fileName === 'index') {
    fileName = path.basename(path.dirname(source));
    parts.pop(); // remove duplicated folder name
  }

  parts.push(`${fileName}${suffix || ''}`);
  return parts.join('/');
}

export function inferTitle(ast: Node): string | undefined {
  let docTitle: string | undefined;
  visit(ast, 'heading', (node) => {
    const { depth } = node;
    if (depth !== 1) return;
    docTitle = toString(node);
  });
  return docTitle;
}

export function parseFrontmatter(source: string, ast: Node): object {
  let result = {};
  visit(ast, 'yaml', (node) => {
    try {
      result = YAML.parse(node.value as string);
    } catch (e) {
      console.error(`Error while parsing frontmatter in ${source}: `, e);
    }
  });

  return result;
}

export function isValidUrl(s: string): boolean {
  try {
    new url.URL(s);
    return true;
  } catch {
    return false;
  }
}

export function isAnchorUrl(s: string): boolean {
  return s[0] === '#';
}
