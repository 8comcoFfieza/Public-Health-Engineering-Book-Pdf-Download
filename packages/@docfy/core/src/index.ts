import glob from 'glob';
import fs from 'fs';
import path from 'path';
import trough from 'trough';
import { Node } from 'unist';
import { Page, Context, Options } from './types';
import { inferTitle, generateUrl, parseFrontmatter } from './utils';
import { createRemark } from './remark';
import { fixUrls, combineDemos, toc } from './plugins';

const DEFAULT_IGNORE = [
  '/**/node_modules/**',
  '/**/.git/**',
  '/**/tmp/**',
  '/**/dist/**',
  'node_modules/**',
  '.git/**',
  'tmp/**',
  'dist/**'
];

function createPage(
  source: string,
  markdown: string,
  ast: Node,
  urlPrefix?: string,
  urlSuffix?: string
): Page {
  const frontmatter = parseFrontmatter(source, ast);

  return {
    source,
    ast,
    markdown,
    metadata: {
      title: inferTitle(ast),
      ...frontmatter,
      url: generateUrl(source, frontmatter, urlPrefix, urlSuffix)
    },
    rendered: ''
  };
}

function initialize(options: Options): Context {
  const ctx: Context = {
    root: options.root,
    remark: createRemark(options.remarkPlugins),
    pages: [],
    settings: {
      tocMaxDepth: options.tocMaxDepth || 6
    }
  };

  options.sources.forEach((item) => {
    const files = glob.sync(item.pattern, {
      root: options.root,
      ignore: [...DEFAULT_IGNORE, ...(item.ignore || [])]
    });

    files.forEach((file) => {
      const relativePath = file.replace(path.join(options.root, '/'), '');
      const markdown = fs.readFileSync(file).toString();
      const ast = ctx.remark.runSync(ctx.remark.parse(markdown));

      ctx.pages.push(
        createPage(relativePath, markdown, ast, item.urlPrefix, item.urlSuffix)
      );
    });
  });

  return ctx;
}

function renderMarkdown(context: Context): void {
  context.pages.forEach((item) => {
    item.rendered = context.remark.stringify(item.ast);
  });
}

export default function (options: Options): Promise<Page[]> {
  return new Promise((resolve, reject) => {
    trough<Context>()
      .use<Options>(initialize)
      .use(combineDemos)
      .use(fixUrls)

      // Make sure TOC and renderMarkdown pluings are the last ones
      .use(toc)
      .use(renderMarkdown)
      .run(options, (err: unknown, ctx: Context): void => {
        if (err) {
          reject(err);
        } else {
          resolve(ctx.pages);
        }
      });
  });
}
