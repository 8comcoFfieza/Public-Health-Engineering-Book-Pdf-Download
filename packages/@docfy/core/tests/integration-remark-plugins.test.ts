import Docfy from '../src';
import { DocfyResult } from '../src/types';
import path from 'path';
import autolinkHeadings from 'remark-autolink-headings';

const root = path.resolve(__dirname, './__fixtures__/monorepo');

describe('When proving remark plugins', () => {
  describe('When plugin has no options', () => {
    let result: DocfyResult;

    beforeAll(async () => {
      const docfy = new Docfy({
        remarkPlugins: [autolinkHeadings]
      });
      result = await docfy.run([
        {
          root,
          urlPrefix: 'docs',
          urlSchema: 'manual',
          pattern: '**/*.md'
        }
      ]);
    });

    test('it should have rendered the content with the plugin with no options', async () => {
      const htmls = [];
      result.content.forEach((page) => {
        htmls.push([page.source, page.rendered]);
      });

      expect(htmls).toMatchSnapshot();
    });
  });

  describe('When plugin has options', () => {
    let result: DocfyResult;

    beforeAll(async () => {
      const docfy = new Docfy({
        remarkPlugins: [[autolinkHeadings, { behavior: 'append' }]]
      });
      result = await docfy.run([
        {
          root,
          urlPrefix: 'docs',
          urlSchema: 'manual',
          pattern: '**/*.md'
        }
      ]);
    });

    test('it should have rendered the content with the plugin and their options', async () => {
      const htmls = [];
      result.content.forEach((page) => {
        htmls.push([page.source, page.rendered]);
      });

      expect(htmls).toMatchSnapshot();
    });
  });
});
