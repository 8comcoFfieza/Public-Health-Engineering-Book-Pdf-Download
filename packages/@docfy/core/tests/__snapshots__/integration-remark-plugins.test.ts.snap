// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When proving remark plugins When plugin has no options it should have rendered the content with the plugin with no options 1`] = `
Array [
  Array [
    "docs/about.md",
    "<h1 id=\\"about-the-project\\"><a href=\\"#about-the-project\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a>About the project</h1>
<h2 id=\\"history\\"><a href=\\"#history\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a>History</h2>
<p>Description.</p>
<h2 id=\\"where-it-came-from\\"><a href=\\"#where-it-came-from\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a>Where it came from</h2>
<p>Description.</p>
<h3 id=\\"but-why\\"><a href=\\"#but-why\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a>But why?</h3>
<p>Description.</p>
<h2 id=\\"another-topic\\"><a href=\\"#another-topic\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a>Another topic</h2>
<h2 id=\\"2-level\\"><a href=\\"#2-level\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a>2 level</h2>
<h3 id=\\"3-level\\"><a href=\\"#3-level\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a>3 level</h3>
<h3 id=\\"3-level-2nd\\"><a href=\\"#3-level-2nd\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a>3 level 2nd</h3>
<h4 id=\\"4-level\\"><a href=\\"#4-level\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a>4 level</h4>
<h5 id=\\"5-level\\"><a href=\\"#5-level\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a>5 level</h5>
<h6 id=\\"6-level\\"><a href=\\"#6-level\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a>6 level</h6>
<h4 id=\\"4-level-2nd\\"><a href=\\"#4-level-2nd\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a>4 level 2nd</h4>
",
  ],
  Array [
    "docs/how-it-works.md",
    "<h1 id=\\"how-it-works\\"><a href=\\"#how-it-works\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a>How it works?</h1>
<p>This is a description.</p>
",
  ],
  Array [
    "packages/package1/components/button.md",
    "<h1 id=\\"button-component\\"><a href=\\"#button-component\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a>Button Component</h1>
",
  ],
  Array [
    "packages/package1/components/form/index.md",
    "<h1 id=\\"form-component\\"><a href=\\"#form-component\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a>Form Component</h1>
",
  ],
  Array [
    "packages/package1/docs/intro.md",
    "<h1 id=\\"docs-intro\\"><a href=\\"#docs-intro\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a>Docs Intro</h1>
",
  ],
  Array [
    "packages/package2/docs/overview.md",
    "<h1 id=\\"this-is-the-overview-page\\"><a href=\\"#this-is-the-overview-page\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a>This is the overview page</h1>
<p>This is a <a href=\\"/docs/package1/components/button\\">link to another file in a different package</a>.</p>
",
  ],
  Array [
    "packages/package2/docs/styles.md",
    "<h1 id=\\"how-styles-work\\"><a href=\\"#how-styles-work\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a>How styles work</h1>
<p>Short description.</p>
<p>Please refer to <a href=\\"/docs/package2/overview\\">overview</a> to get started.</p>
",
  ],
]
`;

exports[`When proving remark plugins When plugin has options it should have rendered the content with the plugin and their options 1`] = `
Array [
  Array [
    "docs/about.md",
    "<h1 id=\\"about-the-project\\">About the project<a href=\\"#about-the-project\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a></h1>
<h2 id=\\"history\\">History<a href=\\"#history\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a></h2>
<p>Description.</p>
<h2 id=\\"where-it-came-from\\">Where it came from<a href=\\"#where-it-came-from\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a></h2>
<p>Description.</p>
<h3 id=\\"but-why\\">But why?<a href=\\"#but-why\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a></h3>
<p>Description.</p>
<h2 id=\\"another-topic\\">Another topic<a href=\\"#another-topic\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a></h2>
<h2 id=\\"2-level\\">2 level<a href=\\"#2-level\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a></h2>
<h3 id=\\"3-level\\">3 level<a href=\\"#3-level\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a></h3>
<h3 id=\\"3-level-2nd\\">3 level 2nd<a href=\\"#3-level-2nd\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a></h3>
<h4 id=\\"4-level\\">4 level<a href=\\"#4-level\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a></h4>
<h5 id=\\"5-level\\">5 level<a href=\\"#5-level\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a></h5>
<h6 id=\\"6-level\\">6 level<a href=\\"#6-level\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a></h6>
<h4 id=\\"4-level-2nd\\">4 level 2nd<a href=\\"#4-level-2nd\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a></h4>
",
  ],
  Array [
    "docs/how-it-works.md",
    "<h1 id=\\"how-it-works\\">How it works?<a href=\\"#how-it-works\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a></h1>
<p>This is a description.</p>
",
  ],
  Array [
    "packages/package1/components/button.md",
    "<h1 id=\\"button-component\\">Button Component<a href=\\"#button-component\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a></h1>
",
  ],
  Array [
    "packages/package1/components/form/index.md",
    "<h1 id=\\"form-component\\">Form Component<a href=\\"#form-component\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a></h1>
",
  ],
  Array [
    "packages/package1/docs/intro.md",
    "<h1 id=\\"docs-intro\\">Docs Intro<a href=\\"#docs-intro\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a></h1>
",
  ],
  Array [
    "packages/package2/docs/overview.md",
    "<h1 id=\\"this-is-the-overview-page\\">This is the overview page<a href=\\"#this-is-the-overview-page\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a></h1>
<p>This is a <a href=\\"/docs/package1/components/button\\">link to another file in a different package</a>.</p>
",
  ],
  Array [
    "packages/package2/docs/styles.md",
    "<h1 id=\\"how-styles-work\\">How styles work<a href=\\"#how-styles-work\\" aria-hidden=\\"true\\" tabindex=\\"-1\\"><span class=\\"icon icon-link\\"></span></a></h1>
<p>Short description.</p>
<p>Please refer to <a href=\\"/docs/package2/overview\\">overview</a> to get started.</p>
",
  ],
]
`;
